Show complaint status and implement sorting in complaint list

rpadovani opened the pull request:
<!-- Thanks for contributing to ArTEMiS! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested the changes and all related features on the test server https://artemistest.ase.in.tum.de.
- [x] I documented my source code using the JavaDoc / JSDoc style.
- [x] ~I added integration test cases for the server (Spring) related to the features~
- [x] ~I added integration test cases for the client (Jest) related to the features~
- [x] I added screenshots/screencast of my UI changes

### Motivation and Context

Now in the complaint list page there is also the status of each complaint, and it is possible to sort the complaint table with different criteria

### Screenshots

![image](https://user-images.githubusercontent.com/1713343/59565268-5982b600-9051-11e9-945f-4230e1fdbc9f.png)

krusche commented:
![image](https://user-images.githubusercontent.com/744067/59565524-0f9bcf00-9055-11e9-8db1-e6d410849f4f.png)
This is how it looks like for tutors (on production). I think there is a problem with the table, because you hide the headers but not the actual rows. Can you please fix it?

krusche commented:
And one more small request: Can we also add a filter to hide the completed ones? In EIST we have now > 200 complaints and sometimes we only want to focus on the ones that are not yet completed, i.e. the status is "No reply yet"

rpadovani commented:
@krusche I added a flag for showing addressed complaints and fixed the visualization. I also forced Spring data to load the assessor, so it should be shown now!

krusche commented:
Did you fix the other points as well?

rpadovani commented:
@krusche sorry, I didn't see the `EDIT`

> * There are way too many objects nested in the complaint. It actually goes until the course. Please remove the participation and the submission from the result as they are not necessary!

We actually need them, to show the exercise title and the submission id

> * I think the assessor is missing (even for me as instructor), but would be interested so that instructors / tutors can contact the assessor to clarify things.

I explicitly added to the query to retrieve assessors as well
 
> * Please double check that the student's identify is never revealed for a tutor (even not in the json response). So the server absolutely has to hide this information when tutors are reviewing the complaints

It seems to me this is already the case

krusche commented:
> @krusche sorry, I didn't see the `EDIT`
> 
> > * There are way too many objects nested in the complaint. It actually goes until the course. Please remove the participation and the submission from the result as they are not necessary!
> 
> We actually need them, to show the exercise title and the submission id
> 

Then please remove everything that is not needed, if we just need the exercise title and the submission id, we can cutoff all other attributes.

> > * I think the assessor is missing (even for me as instructor), but would be interested so that instructors / tutors can contact the assessor to clarify things.
> 
> I explicitly added to the query to retrieve assessors as well

Sounds good. I will test it.
 
> > * Please double check that the student's identify is never revealed for a tutor (even not in the json response). So the server absolutely has to hide this information when tutors are reviewing the complaints
> 
> It seems to me this is already the case

That's not enough. Please test it thoroughly. It would also be nice to have some server and client integration tests to verify this in the future. You actually have immediately strike through the two related bullet points. Could you please add some basic tests for complaints as well? Talk to @thilo-behnke if you have questions about testing

thilo-behnke commented on src/main/webapp/app/list-of-complaints/list-of-complaints.component.ts:
```diff
@@ -55,6 +60,7 @@ export class ListOfComplaintsComponent implements OnInit {
         complaintResponse.subscribe(
             res => {
                 this.complaints = res.body;
+                this.complaintsToShow = this.complaints.filter(c => c.accepted === undefined);
```
If the endpoint response takes a while, it might be that the user has changed the checkbox state for showAddressedComplaints.
Then the complaints are still filtered, even though showAddressedComplaints is true.
I think it would be best to disable the checkbox while loading.

rpadovani replied:
Good one, thanks!

thilo-behnke commented on src/main/webapp/app/list-of-complaints/list-of-complaints.component.ts:
```diff
@@ -96,4 +102,16 @@ export class ListOfComplaintsComponent implements OnInit {
+        if (this.showAddressedComplaints) {
+            this.complaintsToShow = this.complaints;
+        } else {
+            this.complaintsToShow = this.complaints.filter(c => c.accepted === undefined);
```
At first I thought that if the checkbox is activate ALSO the addressed commands are shown.
But after reading this again, this is more of a toggle.
I think it would be more clear if the it was a radio button or a toggle button (https://ng-bootstrap.github.io/#/components/buttons/examples) - show addressed vs. show unaddressed.

thilo-behnke replied:
I was wrong about this - never mind.

thilo-behnke commented on src/main/java/de/tum/in/www1/artemis/repository/ComplaintRepository.java:
```diff
@@ -81,7 +81,7 @@
      * @return a list of complaints
      */
-    @EntityGraph(attributePaths = { "result.participation", "result.submission" })
+    @EntityGraph(attributePaths = { "result.participation", "result.submission", "result.assessor" })
```
Very cool!

krusche commented:
> 1. I don't like that the table is completely hidden, if there are no complaints. Normally, the table should just be empty and show a placeholder entry a la "no results". But I'm not sure how this was discussed with @krusche.

Good comment. If the table is empty, please show a label with something like "No complaints".

krusche commented:
I have one additional feature request. Can we highlight complaints somehow that are not resolved yet and that have been submitted more than one week ago. This makes sure that tutors know which complaints have a higher priority.

Would be great if we can also accomplish this in other places, where complaints are shown.

rpadovani commented:
@thilo-behnke I've tried to write a test, but I have this error:

```
Error: Unexpected value 'undefined' exported by the module 'ArTEMiSSharedModule'
```

Do you know what could be the reason?
I pushed the failing test :)

thilo-behnke commented:
> @thilo-behnke I've tried to write a test, but I have this error:
> 
> ```
> Error: Unexpected value 'undefined' exported by the module 'ArTEMiSSharedModule'
> ```
> 
> Do you know what could be the reason?
> I pushed the failing test :)

@rpadovani Good question 🤔 I can't confirm at the moment, but I would assume that the module `ArTEMiSSharedCommonModule` that is imported and exported by the SharedModule can't be found and also has to be included in the imports of the TestModule. (If this doesn't work, maybe you can narrow the issue down by only importing the explicit pipes / components that you need into the TestComponent).

rpadovani commented:
@krusche @thilo-behnke I think I was able to address all your comments, and I was finally able to setup tests - let me know what you think!

krusche commented on src/main/webapp/app/list-of-complaints/list-of-complaints.component.ts:
```diff
@@ -55,12 +63,14 @@ export class ListOfComplaintsComponent implements OnInit {
         complaintResponse.subscribe(
             res => {
                 this.complaints = res.body!;
+                this.complaintsToShow = this.complaints.filter(c => c.accepted === undefined);
```
```suggestion
                this.complaintsToShow = this.complaints.filter(complaint => complaint.accepted === undefined);
```

krusche commented on src/main/webapp/app/list-of-complaints/list-of-complaints.component.ts:
```diff
@@ -96,4 +106,30 @@ export class ListOfComplaintsComponent implements OnInit {
+        if (this.showAddressedComplaints) {
+            this.complaintsToShow = this.complaints;
+        } else {
+            this.complaintsToShow = this.complaints.filter(c => c.accepted === undefined);
```
```suggestion
            this.complaintsToShow = this.complaints.filter(complaint => complaint.accepted === undefined);
```
